### PR TITLE
OAK-10098 | Support Index purge for ES indexes

### PR DESCRIPTION
**Pull request by nit0906**

### CODE REVIEW

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/IndexConstants.java [hunk 1]
```diff
@@ -27,6 +27,11 @@ public interface IndexConstants {
 
     String TYPE_PROPERTY_NAME = "type";
 
+    /**
+     * Property to hold the value for original index implementation in case the index type is set to disabled.
+     */
+    String ORIGINAL_TYPE_PROPERTY_NAME = "orig_type";
```
Comment (fabriziofortino) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/IndexConstants.java]: two questions:
* should we use a hidden property here prefixing it with `:`?
* I would use the camel case notation to be consistent with the other constants (eg: `originalType` or `origType`)

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java [hunk 2]
```diff
@@ -90,25 +93,28 @@ public static List<IndexVersionOperation> generateIndexVersionOperationList(Node
...
         NodeState indexDefParentNode = NodeStateUtils.getNode(rootNode, parentPath);
         List<IndexName> reverseSortedIndexNameList = getReverseSortedIndexNameList(indexNameObjectList);
         List<IndexVersionOperation> indexVersionOperationList = new LinkedList<>();
 
         List<IndexName> disableIndexNameObjectList = removeDisabledCustomIndexesFromList(indexDefParentNode, reverseSortedIndexNameList);
         // for disabled indexes, we check if they exist in read-only repo, it not anymore, do full deletion then, otherwise, no action needed
+        // Also skip deletion if ignoreIsIndexActiveCheck is true (which would be from ElasticPurgeOldIndexVersion) - this would be for ES indexes where we don't have any way to determine if the disabled ES index is
+        // So we skip deleting the disabled index here so that we don't delete a disabled base index by mistake.
+        // Summary - Delete disabled index if ignoreIsIndexActiveCheck = false (which would be when called from LucenePurgeOldIndexVersion) and hidden oak mount doesn't exists.
```
Comment (fabriziofortino) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java]: typo
```suggestion
        // Summary - Delete disabled index if ignoreIsIndexActiveCheck = false (which would be when called from LucenePurgeOldIndexVersion) and hidden oak mount doesn't exist.
```

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java [hunk 3]
```diff
@@ -90,25 +93,28 @@ public static List<IndexVersionOperation> generateIndexVersionOperationList(Node
...
     public static List<IndexVersionOperation> generateIndexVersionOperationList(NodeState rootNode, String parentPath,
-            List<IndexName> indexNameObjectList, long purgeThresholdMillis, boolean shouldPurgeBaseIndex) {
+            List<IndexName> indexNameObjectList, long purgeThresholdMillis, boolean shouldPurgeBaseIndex, boolean ignoreIsIndexActiveCheck) {
         NodeState indexDefParentNode = NodeStateUtils.getNode(rootNode, parentPath);
         List<IndexName> reverseSortedIndexNameList = getReverseSortedIndexNameList(indexNameObjectList);
         List<IndexVersionOperation> indexVersionOperationList = new LinkedList<>();
 
         List<IndexName> disableIndexNameObjectList = removeDisabledCustomIndexesFromList(indexDefParentNode, reverseSortedIndexNameList);
         // for disabled indexes, we check if they exist in read-only repo, it not anymore, do full deletion then, otherwise, no action needed
+        // Also skip deletion if ignoreIsIndexActiveCheck is true (which would be from ElasticPurgeOldIndexVersion) - this would be for ES indexes where we don't have any way to determine if the disabled ES index is
```
Comment (fabriziofortino) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java]: could you better describe this part?
`this would be for ES indexes where we don't have any way to determine if the disabled ES index is`

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/PurgeOldIndexVersion.java [hunk 4]
```diff
@@ -166,13 +175,39 @@ private Iterable<String> getRepositoryIndexPaths(NodeStore store) throws CommitF
...
-    private Set<String> filterIndexPaths(Iterable<String> repositoryIndexPaths, List<String> commandlineIndexPaths) {
+    private Set<String> filterIndexPaths(NodeStore store, Iterable<String> repositoryIndexPaths, List<String> commandlineIndexPaths) {
         Set<String> filteredIndexPaths = new HashSet<>();
         for (String commandlineIndexPath : commandlineIndexPaths) {
             for (String repositoryIndexPath : repositoryIndexPaths) {
                 if (PurgeOldVersionUtils.isIndexChildNode(commandlineIndexPath, repositoryIndexPath)
                         || PurgeOldVersionUtils.isBaseIndexEqual(commandlineIndexPath, repositoryIndexPath)) {
-                    filteredIndexPaths.add(repositoryIndexPath);
+                    NodeState root = store.getRoot();
+                    NodeBuilder rootBuilder = root.builder();
```
Comment (fabriziofortino) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/PurgeOldIndexVersion.java]: can we move these two statements outside the nested fors

File: oak-run-elastic/src/main/java/org/apache/jackrabbit/oak/indexversion/ElasticPurgeOldIndexVersion.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        } else {
+            coordinate = buildStep.build();
+        }
+
+        closer.register(coordinate);
+
+        String remoteIndexName = ElasticIndexNameHelper.getRemoteIndexName(indexPrefix, idxPath, SEED_VALUE);
+
+        ElasticsearchIndicesClient client = coordinate.getClient().indices();
+        DeleteIndexResponse deleteIndexResponse = null;
```
Comment (fabriziofortino) [inline oak-run-elastic/src/main/java/org/apache/jackrabbit/oak/indexversion/ElasticPurgeOldIndexVersion.java]:
```suggestion
        DeleteIndexResponse deleteIndexResponse;
```

File: oak-run-elastic/src/test/java/org/apache/jackrabbit/oak/index/ElasticRepositoryFixture.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+package org.apache.jackrabbit.oak.index;
+
+import org.apache.jackrabbit.oak.Oak;
+import org.apache.jackrabbit.oak.plugins.index.elastic.ElasticConnection;
+import org.apache.jackrabbit.oak.plugins.index.elastic.ElasticIndexTracker;
+import org.apache.jackrabbit.oak.plugins.index.elastic.ElasticMetricHandler;
+import org.apache.jackrabbit.oak.plugins.index.elastic.index.ElasticIndexEditorProvider;
+import org.apache.jackrabbit.oak.plugins.index.elastic.query.ElasticIndexProvider;
+import org.apache.jackrabbit.oak.plugins.index.search.ExtractedTextCache;
+import org.apache.jackrabbit.oak.spi.commit.Observer;
```
Comment (fabriziofortino) [inline oak-run-elastic/src/test/java/org/apache/jackrabbit/oak/index/ElasticRepositoryFixture.java]: unused
```suggestion
```

File: oak-run-elastic/src/test/java/org/apache/jackrabbit/oak/index/ElasticRepositoryFixture.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+    public ElasticRepositoryFixture(File dir, ElasticConnection connection) {
+        super(dir);
+        this.connection = connection;
+    }
+
+    public ElasticRepositoryFixture(File dir, NodeStore nodeStore, ElasticConnection connection) {
+        super(dir, nodeStore);
+        this.connection = connection;
+    }
```
Comment (fabriziofortino) [inline oak-run-elastic/src/test/java/org/apache/jackrabbit/oak/index/ElasticRepositoryFixture.java]: never invoked, can we remove it?

File: oak-run-elastic/src/test/java/org/apache/jackrabbit/oak/index/ElasticRepositoryFixture.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Override
+    protected void configureIndexProvider(Oak oak) throws IOException {
```
Comment (fabriziofortino) [inline oak-run-elastic/src/test/java/org/apache/jackrabbit/oak/index/ElasticRepositoryFixture.java]: throws can be removed
```suggestion
    protected void configureIndexProvider(Oak oak) {
```

File: oak-run-elastic/src/test/java/org/apache/jackrabbit/oak/index/ElasticRepositoryFixture.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        ElasticIndexTracker tracker = new ElasticIndexTracker(connection, new ElasticMetricHandler(StatisticsProvider.NOOP));
+
+        ElasticIndexEditorProvider ep = new ElasticIndexEditorProvider(tracker, connection,new ExtractedTextCache(0,0));
+        ElasticIndexProvider provider = new ElasticIndexProvider(tracker);
+        oak.with((QueryIndexProvider) provider)
```
Comment (fabriziofortino) [inline oak-run-elastic/src/test/java/org/apache/jackrabbit/oak/index/ElasticRepositoryFixture.java]: we should not need to cast it
```suggestion
        oak.with(provider)
```

File: oak-run/src/test/java/org/apache/jackrabbit/oak/index/LuceneRepositoryFixture.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    public LuceneRepositoryFixture(File dir) {
+        super(dir);
+    }
+
+    public LuceneRepositoryFixture(File dir, NodeStore nodeStore) {
+        super(dir, nodeStore);
+    }
+
+    @Override
+    protected void configureIndexProvider(Oak oak) throws IOException {
```
Comment (fabriziofortino) [inline oak-run/src/test/java/org/apache/jackrabbit/oak/index/LuceneRepositoryFixture.java]: throws can be removed
```suggestion
    protected void configureIndexProvider(Oak oak) {
```

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/IndexConstants.java (hunk 1 above)
Comment (nit0906) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/IndexConstants.java]: Yes - hidden property might make more sense

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java (hunk 3 above)
Comment (nit0906) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java]: @fabriziofortino  - done. Can you  recheck once if the new comment is better ?

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java (hunk 2 above)
Comment (nit0906) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java]: I think exists is correct since oak mount is singular and third person.

**Review by fabriziofortino — APPROVED** [review]

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java [hunk 11]
```diff
@@ -75,7 +78,7 @@ public String toString() {
      */
     public static List<IndexVersionOperation> generateIndexVersionOperationList(NodeState rootNode, String parentPath,
                                                                                 List<IndexName> indexNameObjectList, long purgeThresholdMillis) {
-        return generateIndexVersionOperationList(rootNode, parentPath, indexNameObjectList, purgeThresholdMillis, true);
+        return generateIndexVersionOperationList(rootNode, parentPath, indexNameObjectList, purgeThresholdMillis, true, false);
```
Comment (tihom88) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java]: Shouldn't there be two different implementations for lucene and elastic (different classes).

I think we already have an issue to have similar hidden mounts for elastic index, to make elastic compatible with blue-green deployment

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/PurgeOldIndexVersion.java [hunk 12]
```diff
@@ -166,13 +175,38 @@ private Iterable<String> getRepositoryIndexPaths(NodeStore store) throws CommitF
...
+
+                        // If type = disabled and orig_type is not set - this means someone manually marked the index as disabled
+                        // We can't determine the original index implementation in this case - so we let it run for whichever PurgeOldIndexVersion's impl calls it first
+                        // Two scenarios to note here -
+                        // 1. ElasticPurgeOldIndexVersion calls this for a lucene index (deletes the index from oak repo - which we want)
+                        // and then in postDeleteOp tries to delete a non existent remote elastic index - that particular call will simply fail.
+                        // 2. LucenePurgeOldIndexVersion calls this for an elastic index (deletes the index from oak repo)
+                        // and then does a NOOP in postDeleteOp - effectively we are left with a dangling elastic remote index in this case.
+                        filteredIndexPaths.add(repositoryIndexPath);
+                    }
```
Comment (tihom88) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/PurgeOldIndexVersion.java:208]: I think we can get solve dangling elastic issue in following manner.
In case of disabled indexes both elastic and lucene purge operations should evaluate and add property in hidden node mentioning the state wrt to both elastic and lucene purge operations.
e.g. purgeLucene = true and purgeElastic=true on :purge node.

and postDeleteOp we check if both these properties are true we delete the index, else this index will remain in disabled state.

Comment (nit0906) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/PurgeOldIndexVersion.java:209]: @tihom88  - changes as part of this PR are already adding a hidden property on the index :originalType in case this particular code marks it as disabled. That case is already handled above and would not lead to a dangling ES index. 

So basically - if :originalType is elasticsearch and type=disabled and this code gets called from ES impl - then the ES index from remote will also get deleted.

However if someone manually had marked the index as disabled - then of course there's no way to determine the original index type (unless someone also sets the :originalType property in that case). - This is the only case where there would be a dangling index left.

Or were you suggesting something else ?

Comment (nit0906) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/PurgeOldIndexVersion.java:208]: @tihom88  - made the changes to actually skip disabled indexes in case of type=disabled and :originalType prop not set. This would basically exclude manually disabled indexes from purge automated logic. Seems a bit more clean this way.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java (hunk 11 above)
Comment (nit0906) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java]: Made the changes to make it into 2 separate implementations so that the logic is a bit more clean

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/PurgeOldIndexVersion.java [hunk 13]
```diff
@@ -195,21 +227,51 @@ private void purgeOldIndexVersion(NodeStore store, List<IndexVersionOperation> t
...
             } else {
                 LOG.error("nodebuilder null for path " + nodeName);
             }
         }
     }
+
+    /**
+     * @return IndexType served by the implementation like lucene or elasticsearch
+     */
+    protected abstract String getIndexType();
```
Comment (rishabhdaim) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/PurgeOldIndexVersion.java:260]: Since we are using `getIndexType()` without any null check, I think annotating this method with `@NonNull` is better.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/run/PurgeOldIndexVersionCommand.java [hunk 14]
```diff
@@ -28,11 +29,10 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import joptsimple.OptionParser;
-import joptsimple.OptionSet;
-import joptsimple.OptionSpec;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
```
Comment (rishabhdaim) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/run/PurgeOldIndexVersionCommand.java:33]: Could you please revert the unintended changes in file imports.

File: oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/AbstractIndexTestCommand.java [hunk 15]
```diff
@@ -16,30 +16,28 @@
...
 import javax.jcr.RepositoryException;
 import javax.jcr.Session;
-
-import org.apache.jackrabbit.oak.plugins.index.IndexConstants;
-import org.apache.jackrabbit.oak.plugins.index.lucene.util.LuceneIndexDefinitionBuilder;
-import org.junit.After;
-import org.junit.Rule;
-import org.junit.rules.TemporaryFolder;
+import java.io.File;
+import java.io.IOException;
```
Comment (rishabhdaim) [inline oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/AbstractIndexTestCommand.java:31]: Same as the above comment.

File: oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/IndexRepositoryFixture.java [hunk 16]
```diff
@@ -16,48 +16,42 @@
...
 
+import javax.jcr.Repository;
+import javax.jcr.RepositoryException;
+import javax.jcr.Session;
+import javax.jcr.SimpleCredentials;
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.util.function.Predicate;
+
```
Comment (rishabhdaim) [inline oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/IndexRepositoryFixture.java:40]: Please remove the un-necessary import changes.

File: oak-run-elastic/src/main/java/org/apache/jackrabbit/oak/indexversion/ElasticIndexVersionOperation.java [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+        return false;
+    }
+
+    @Override
+    protected IndexName getActiveIndex(List<IndexName> reverseSortedIndexNameList, String parentPath, NodeState rootNode) {
+        // ES doesn't have a concept of active/inactive index as of now (no hidden oak mount to detect this)
+        // So here we simply return the highest versioned index
+        IndexName indexNameObject = reverseSortedIndexNameList.get(0);
+        reverseSortedIndexNameList.remove(0);
+        return indexNameObject;
```
Comment (rishabhdaim) [inline oak-run-elastic/src/main/java/org/apache/jackrabbit/oak/indexversion/ElasticIndexVersionOperation.java:50]: Since `List.remove(index)` return the element, the above code can be modified as:
```suggestion
        return reverseSortedIndexNameList.remove(0);
```

Comment (rishabhdaim) [inline oak-run-elastic/src/main/java/org/apache/jackrabbit/oak/indexversion/ElasticIndexVersionOperation.java]: Since `List.remove(index)` returns the element at the given index as well, the above code can be modified as:
```suggestion
        return reverseSortedIndexNameList.remove(0);
```

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java [hunk 18]
```diff
@@ -254,6 +242,24 @@ enum Operation {
...
+    protected abstract boolean checkIfDisabledIndexCanBeMarkedForDeletion(NodeState indexNode);
+
+    /**
+     *
+     * @param reverseSortedIndexNameList
+     * @param parentPath
+     * @param rootNode
+     * @return Highest versioned active index's IndexName
+     */
+    protected abstract IndexName getActiveIndex(List<IndexName> reverseSortedIndexNameList, String parentPath, NodeState rootNode);
```
Comment (rishabhdaim) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java:267]: This method should be annotated with `@Nullable` since it can return a null value.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/run/PurgeOldIndexVersionCommand.java [hunk 19]
```diff
@@ -1,25 +1,26 @@
...
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
```
Comment (rishabhdaim) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/run/PurgeOldIndexVersionCommand.java:17]: Please revert these changes.

Comment (nit0906) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/run/PurgeOldIndexVersionCommand.java:17]: @rishabhdaim  - not getting you - what changes do you want me to revert ? this is a license header which is needed in every file. Auto formatting fixed a whitespace in it.

File: oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/IndexRepositoryFixture.java (hunk 16 above)
Comment (nit0906) [reply oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/IndexRepositoryFixture.java:40]: not following - the effective changes here are removal of certain imports like 

import org.apache.jackrabbit.oak.plugins.index.lucene.LuceneIndexEditorProvider;
import org.apache.jackrabbit.oak.plugins.index.lucene.LuceneIndexProvider;
import org.apache.jackrabbit.oak.spi.commit.Observer;
import org.apache.jackrabbit.oak.spi.query.QueryIndexProvider;

These are no longer needed in this class.

File: oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/AbstractIndexTestCommand.java (hunk 15 above)
Comment (nit0906) [reply oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/AbstractIndexTestCommand.java:31]: @rishabhdaim  -if you are referring to order of imports - that probably got changed by IDE formatting . I don't think it's doing any harm.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/run/PurgeOldIndexVersionCommand.java (hunk 14 above)
Comment (nit0906) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/run/PurgeOldIndexVersionCommand.java:33]: is there anything apart from the order changes that you see problematic ?

File: oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/AbstractIndexTestCommand.java (hunk 15 above)
Comment (rishabhdaim) [reply oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/AbstractIndexTestCommand.java:31]: Yes, these changes unnecessarily cause distractions to the actual code changes and also increase the change set.

This sometimes causes issues while backporting the changes.

So, I would keep the change set to bare essential changes.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java (hunk 18 above)
Comment (nit0906) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java:267]: ack.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/PurgeOldIndexVersion.java (hunk 13 above)
Comment (nit0906) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/indexversion/PurgeOldIndexVersion.java:260]: ack.

File: oak-run-elastic/src/main/java/org/apache/jackrabbit/oak/indexversion/ElasticIndexVersionOperation.java (hunk 17 above)
Comment (nit0906) [reply oak-run-elastic/src/main/java/org/apache/jackrabbit/oak/indexversion/ElasticIndexVersionOperation.java]: ack
